File: src/hamster_tracker/vision/marker_detector.py
from dataclasses import dataclass
from typing import Optional, Tuple

from .geometry import WheelGeometry


@dataclass(frozen=True)
class MarkerDetection:
    x_px: float
    y_px: float
    area_px: float
    radial_distance_px: float
    score: float


@dataclass(frozen=True)
class MarkerDetectorConfig:
    hsv_lower: Tuple[int, int, int]
    hsv_upper: Tuple[int, int, int]
    expected_radius_px: float
    radius_tolerance_px: float
    min_area_px: float = 30.0
    max_area_px: float = 5000.0
    morphology_kernel: int = 3


class MarkerDetector:
    """HSV blob detector constrained to the marker's expected wheel annulus."""

    def __init__(self, geometry: WheelGeometry, config: MarkerDetectorConfig):
        self.geometry = geometry
        self.config = config

    def detect(self, bgr_frame) -> Optional[MarkerDetection]:
        try:
            import cv2
            import numpy as np
        except ImportError as exc:  # pragma: no cover - depends on target environment
            raise RuntimeError(
                "MarkerDetector requires OpenCV and NumPy. Install the 'vision' extra "
                "for desktop development, or use the Jetson system OpenCV build."
            ) from exc

        hsv = cv2.cvtColor(bgr_frame, cv2.COLOR_BGR2HSV)
        lower = np.array(self.config.hsv_lower, dtype=np.uint8)
        upper = np.array(self.config.hsv_upper, dtype=np.uint8)
        mask = cv2.inRange(hsv, lower, upper)

        kernel_size = max(1, int(self.config.morphology_kernel))
        if kernel_size > 1:
            kernel = np.ones((kernel_size, kernel_size), dtype=np.uint8)
            mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
            mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)

        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        best = None

        for contour in contours:
            area = float(cv2.contourArea(contour))
            if area < self.config.min_area_px or area > self.config.max_area_px:
                continue

            moments = cv2.moments(contour)
            if moments["m00"] == 0:
                continue

            x = float(moments["m10"] / moments["m00"])
            y = float(moments["m01"] / moments["m00"])
            radial = self.geometry.radial_distance_px(x, y)
            radial_error = abs(radial - self.config.expected_radius_px)
            if radial_error > self.config.radius_tolerance_px:
                continue

            radial_score = 1.0 - radial_error / max(self.config.radius_tolerance_px, 1e-6)
            area_score = min(1.0, area / max(self.config.min_area_px * 4.0, 1.0))
            score = 0.8 * radial_score + 0.2 * area_score

            detection = MarkerDetection(x, y, area, radial, score)
            if best is None or detection.score > best.score:
                best = detection

        return best

File: android/app/src/main/java/com/hujunhan/hamsterwheeltracker/vision/MarkerDetector.kt
package com.hujunhan.hamsterwheeltracker.vision

import org.opencv.core.Core
import org.opencv.core.Mat
import org.opencv.core.MatOfPoint
import org.opencv.core.Scalar
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import kotlin.math.hypot

/** Android/OpenCV port of the Python reference HSV + annulus detector. */
class MarkerDetector {
    private val rgb = Mat()
    private val hsv = Mat()
    private val mask = Mat()
    private val hierarchy = Mat()
    private var morphologyKernel = Mat()
    private var morphologyKernelSize = 0

    fun detect(rgbaFrame: Mat, config: CalibrationConfig): MarkerFrameResult {
        val resolved = config.resolved(rgbaFrame.cols(), rgbaFrame.rows())

        Imgproc.cvtColor(rgbaFrame, rgb, Imgproc.COLOR_RGBA2RGB)
        Imgproc.cvtColor(rgb, hsv, Imgproc.COLOR_RGB2HSV)
        Core.inRange(
            hsv,
            Scalar(config.hsvLowerH.toDouble(), config.hsvLowerS.toDouble(), config.hsvLowerV.toDouble()),
            Scalar(config.hsvUpperH.toDouble(), 255.0, 255.0),
            mask,
        )

        val kernelSize = config.morphologyKernel.coerceAtLeast(1)
        if (kernelSize > 1) {
            ensureMorphologyKernel(kernelSize)
            Imgproc.morphologyEx(mask, mask, Imgproc.MORPH_OPEN, morphologyKernel)
            Imgproc.morphologyEx(mask, mask, Imgproc.MORPH_CLOSE, morphologyKernel)
        }

        val contours = mutableListOf<MatOfPoint>()
        val candidates = mutableListOf<MarkerCandidate>()
        var best: MarkerDetection? = null

        Imgproc.findContours(mask, contours, hierarchy, Imgproc.RETR_EXTERNAL, Imgproc.CHAIN_APPROX_SIMPLE)
        try {
            for (contour in contours) {
                val area = Imgproc.contourArea(contour)
                if (area < config.minAreaPx || area > config.maxAreaPx) {
                    val rect = Imgproc.boundingRect(contour)
                    candidates += MarkerCandidate(
                        xPx = rect.x + rect.width / 2f,
                        yPx = rect.y + rect.height / 2f,
                        areaPx = area,
                        accepted = false,
                        rejection = "area",
                    )
                    continue
                }

                val moments = Imgproc.moments(contour)
                if (moments.m00 == 0.0) continue

                val x = (moments.m10 / moments.m00).toFloat()
                val y = (moments.m01 / moments.m00).toFloat()
                val radial = hypot(
                    (x - resolved.centerX).toDouble(),
                    (y - resolved.centerY).toDouble(),
                ).toFloat()
                val radialError = kotlin.math.abs(radial - resolved.expectedMarkerRadiusPx)
                if (radialError > resolved.radiusTolerancePx) {
                    candidates += MarkerCandidate(x, y, area, false, "annulus")
                    continue
                }

                val radialScore = 1.0 - radialError / resolved.radiusTolerancePx.coerceAtLeast(1e-6f)
                val areaScore = (area / (config.minAreaPx * 4.0).coerceAtLeast(1.0)).coerceAtMost(1.0)
                val score = (0.8 * radialScore + 0.2 * areaScore).toFloat()
                val detection = MarkerDetection(x, y, area, radial, score)
                candidates += MarkerCandidate(x, y, area, true, null)
                if (best == null || detection.score > best!!.score) best = detection
            }
        } finally {
            contours.forEach { it.release() }
        }

        return MarkerFrameResult(
            resolvedCalibration = resolved,
            detection = best,
            candidates = candidates.take(MAX_DEBUG_CANDIDATES),
        )
    }

    fun hsvAt(xPx: Float, yPx: Float): HsvSample? {
        if (hsv.empty()) return null
        val x = xPx.toInt().coerceIn(0, hsv.cols() - 1)
        val y = yPx.toInt().coerceIn(0, hsv.rows() - 1)
        val values = hsv.get(y, x) ?: return null
        if (values.size < 3) return null
        return HsvSample(values[0].toInt(), values[1].toInt(), values[2].toInt())
    }

    fun close() {
        rgb.release()
        hsv.release()
        mask.release()
        hierarchy.release()
        morphologyKernel.release()
    }

    private fun ensureMorphologyKernel(size: Int) {
        if (size == morphologyKernelSize && !morphologyKernel.empty()) return
        morphologyKernel.release()
        morphologyKernel = Imgproc.getStructuringElement(
            Imgproc.MORPH_RECT,
            Size(size.toDouble(), size.toDouble()),
        )
        morphologyKernelSize = size
    }

    companion object {
        private const val MAX_DEBUG_CANDIDATES = 32
    }
}

data class MarkerDetection(
    val xPx: Float,
    val yPx: Float,
    val areaPx: Double,
    val radialDistancePx: Float,
    val score: Float,
)

data class MarkerCandidate(
    val xPx: Float,
    val yPx: Float,
    val areaPx: Double,
    val accepted: Boolean,
    val rejection: String?,
)

data class MarkerFrameResult(
    val resolvedCalibration: ResolvedCalibration,
    val detection: MarkerDetection?,
    val candidates: List<MarkerCandidate>,
    val frameWidth: Int = resolvedCalibration.frameWidth,
    val frameHeight: Int = resolvedCalibration.frameHeight,
    val rotationDegrees: Int = 0,
) {
    fun withRotation(rotationDegrees: Int): MarkerFrameResult = copy(rotationDegrees = rotationDegrees)
}
